### PR TITLE
Add visualizer

### LINKED ISSUE #720: Need to visualize forecast results
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Need to visualize forecast results

GUI needs to get the result image

### Use case

```
yolo.det.visualize
yolo.seg.visualize(img, result)

yolo.seg.visualize
yolo.seg.visualize(img, result)
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by AyushExel**
Fixes - https://github.com/ultralytics/ultralytics/issues/720
```
    from ultralytics import YOLO
    from ultralytics.yolo.utils import ROOT
    import cv2

    model = YOLO("yolov8n-seg.pt")
    source = str(ROOT / "assets/bus.jpg")

    res = model(img_cv)
    resimg = res[0].visualize(show_conf=False)

    cv2.imshow("res", resimg)
    cv2.waitKey(0)

```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Visualize detection results with customizable annotations directly in Ultralytics models.

### 📊 Key Changes
- Added `visualize()` function to `Result` object for plotting detections, including boxes, masks, and class probabilities.
- Extended `Results` class with optional `orig_img` for visualization reference and `names` for class labels.
- Enhanced `Annotator` to support both PIL and OpenCV visualization methods.
- Modified prediction-related files to utilize new visualization capabilities and pass necessary information.

### 🎯 Purpose & Impact
- Allows users to easily visualize model predictions, improving understanding and analysis of detection results. 🖼️
- Streamlines the process of plotting annotations on images, beneficial for both debugging and presenting results elegantly. 📈
- Offers customization options like confidence display and line thickness, providing flexibility for various use cases. 🛠️

### CODE REVIEW

File: ultralytics/yolo/utils/visualizer.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        print("box")
+        cls, conf = d.cls.squeeze(), d.conf.squeeze()
+        c = int(cls)
+        label = f'{names[int(cls)]}' + (f'{conf:.2f}' if show_conf else '')
+        annotator.box_label(d.xyxy.squeeze(), label, color=colors(c, True))
+    
+    annotator.masks(
+            masks.data,
+            colors=[colors(x, True) for x in boxes.cls],
+            im_gpu=torch.as_tensor(img, dtype=torch.float16).permute(2, 0, 1).flip(0).contiguous() /
```
Comment (AyushExel) [inline ultralytics/yolo/utils/visualizer.py]: @Laughing-q can you tell me about the im_gpu argument here? Here this should work the same as retina_masks but it's throwing error.

Comment (Laughing-q) [reply ultralytics/yolo/utils/visualizer.py]: @AyushExel I'm testing it.

File: ultralytics/yolo/utils/visualizer.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    for d in reversed(boxes):
+        print("box")
+        cls, conf = d.cls.squeeze(), d.conf.squeeze()
+        c = int(cls)
+        label = f'{names[int(cls)]}' + (f'{conf:.2f}' if show_conf else '')
+        annotator.box_label(d.xyxy.squeeze(), label, color=colors(c, True))
+
+    annotator.masks(masks.data,
+                    colors=[colors(x, True) for x in boxes.cls],
+                    im_gpu=torch.as_tensor(img, dtype=torch.float16).permute(2, 0, 1).flip(0).contiguous() / 255)
```
Comment (Laughing-q) [inline ultralytics/yolo/utils/visualizer.py:50]: put the img to gpu

File: ultralytics/yolo/utils/visualizer.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+if __name__ == "__main__":
+    from ultralytics import YOLO
+    from ultralytics.yolo.utils import ROOT
+    import cv2
+
+    model = YOLO("yolov8n-seg.pt")
+    img = cv2.imread(str(ROOT / "assets/bus.jpg"))
+    res = model(img)
```
Comment (Laughing-q) [inline ultralytics/yolo/utils/visualizer.py:61]: set `retina_masks=True`

Comment (AyushExel) [reply ultralytics/yolo/utils/visualizer.py]: I think we should support automatic detection of size as most users might not know about retina_masks. I've made an update for that

File: ultralytics/yolo/utils/visualizer.py (hunk 2 above)
Comment (AyushExel) [reply ultralytics/yolo/utils/visualizer.py]: okay got it

**Comment by mahendra-gehlot** [comment]
@AyushExel,  May you add Eigen-Cam for Yolov8? Here is [YOLOv5 ](https://github.com/jacobgil/pytorch-grad-cam/blob/master/tutorials/EigenCAM%20for%20YOLO5.ipynb)implementation of Eigen-CAM by @jacobgil .

**Comment by AyushExel** [comment]
@mahendra-gehlot no plans as of yet. But it looks interesting

File: ultralytics/yolo/engine/results.py [hunk 4]
```diff
@@ -118,6 +123,53 @@ def __getattr__(self, attr):
...
+            img = np.asarray(img)[:, :, ::-1]
+            img = np.ascontiguousarray(img)
+
+        annotator = Annotator(img, line_width, font_size, font, pil, example)
+        boxes = self.boxes
+        masks = self.masks
+        logits = self.probs
+        names = self.names
+
+        if boxes is not None:
```
Comment (Laughing-q) [inline ultralytics/yolo/engine/results.py:154]: I think we should split these functionalities out to `visualize_box`, `visualize_mask`, visualize_prob`. Then users can visualize anything of these boxes, masks, probs as they want.

Comment (AyushExel) [reply ultralytics/yolo/engine/results.py:144]: Yeah that's the plat. Visualize() will visualize everything, visualize_box will only visualize box, and so on

**Comment by AyushExel** [comment]
@Laughing-q hey btw I fixed the callback issue and updated the usage as we discussed. Can you take a look again whenever you're free?

**Comment by Laughing-q** [comment]
@AyushExel sure!

**Comment by Laughing-q** [comment]
@AyushExel all the `Results(orig_shape=self.orig_shape)` should update to `Results(orig_img=self.orig_img)` I think. [https://github.com/ultralytics/ultralytics/blob/41c7e894431fe8962cf6f278c36757f9e6626c37/ultralytics/yolo/engine/results.py#L62](https://github.com/ultralytics/ultralytics/blob/41c7e894431fe8962cf6f278c36757f9e6626c37/ultralytics/yolo/engine/results.py#L62)
And how about moving `names` to `def visualize(...name=None)` instead of making it as a member of `Results`? since `names` is only needed by this visualize.

**Comment by Laughing-q** [comment]
@AyushExel also we probably need more Results-related operation tests in our test sctipt, like `results.cpu()`, `results.numpy()`

**Comment by AyushExel** [comment]
@Laughing-q @glenn-jocher  Added a test and changed the Results arguments

**Comment by glenn-jocher** [comment]
@AyushExel is this good to merge?

**Comment by AyushExel** [comment]
@glenn-jocher yes. We intend to add some more visualizer features later on in separate PRs

**Comment by nikbobo** [comment]
Seens it not work on the latest version, did you remove this function?

**Comment by AyushExel** [comment]
@nikbobo please see the docs here https://docs.ultralytics.com/predict/#plotting-results

**Comment by AyushExel** [comment]
Yeah it should plot everything from box and mask. We just changed the name from visualizer to plot
